File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <string>
#include <vector>
using namespace std;

vector<int> adj[204];
int visited[204];
void dfs(int here){
    visited[here] = 1;
    for(int there: adj[here]){
        if(visited[there])continue;
        dfs(there);
    }
}
int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    for(int i = 0; i < computers.size(); i++){
        for(int j = 0; j < computers[i].size(); j++){
            if(i == j)continue;
            if(computers[i][j]){// 연결
                adj[i].push_back(j);
                adj[j].push_back(i);
            }
        }
    }
    for(int i = 0; i < n; i++){
        if(visited[i])continue;
        dfs(i);
        answer++;
    }
    return answer;
}

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    static boolean[] visited;
    static ArrayList<Integer>[] adj;
    
    public int solution(int n, int[][] computers) {
        int answer = 0;
        visited = new boolean[n + 1];
        adj = new ArrayList[n + 1];
        for(int i = 0; i < n + 1; i++){
            adj[i] = new ArrayList<>();
        }
        
        for(int i = 0; i < computers.length; i++){//0, 1, 2, 3, ...
            for(int j = 0; j < computers[i].length; j++){
                if(i == j)continue;
                if(computers[i][j] != 0)adj[i].add(j);//0, 1, 2, 3, 
            }
        }
        
        for(int i = 0; i < n; i++){
            if(!visited[i]){
                answer++;
                dfs(i);
            }
            
        }
        for(int i = 0; i < n + 1; i++){
            System.out.print(i + " ");
            for(int j : adj[i]){
                System.out.print(j + " ");
            }
            System.out.println();
        }
        return answer;
    }
    
    static void dfs(int here){
        visited[here] = true;
        for(int there : adj[here]){
            if(visited[there])continue;
            dfs(there);
        }
        return;
    }
}